### PR TITLE
Adding all unit tests for provision

### PR DESCRIPTION
**Pull request by johnnyman727**
@rwaldron can you look this over?

### CODE REVIEW

**Comment by johnnyman727** [comment]
@rwaldron This was probably the most painful rebase ever. Could you look it over for me?

**Comment by rwaldron** [comment]
```
rwaldron at nova in ~
$ t2 provision
INFO Connecting to Tessel...
INFO Connected to arnold over USB
INFO Creating public and private keys for Tessel authentication...
INFO SSH Keys written.
INFO Authenticating Tessel with public key...
INFO Tessel authenticated with public key.
/Users/rwaldron/clonez/t2-cli/node_modules/usb/usb.js:42
	this.__close()
	     ^
Error: Can't close device with a pending request
    at Error (native)
    at Device.usb.Device.close (/Users/rwaldron/clonez/t2-cli/node_modules/usb/usb.js:42:7)
    at USBConnection.<anonymous> (/Users/rwaldron/clonez/t2-cli/lib/usb_connection.js:151:17)
    at Device.<anonymous> (/Users/rwaldron/clonez/t2-cli/node_modules/usb/usb.js:185:7)
```

File: lib/tessel/provision.js [hunk 1]
```diff
@@ -8,17 +8,18 @@ var async = require('async');
 var Tessel = require('./tessel');
 var commands = require('./commands');
 var logs = require('../logs');
-var dir = path.join(osenv.home(), '.tessel');
+var directory = osenv.home() + '/.tessel/';
```
Comment (rwaldron) [inline lib/tessel/provision.js]: `path.join(osenv.home(), '.tessel');`

**Comment by johnnyman727** [comment]
@rwaldron thanks for pointing that out - this is a new bug that cropped up because I've started being vigilant about closing the connection once the task is done.

[Partially talking out loud for my own understanding]
The issue here is that the `USBConnection` inherits from `DuplexStream` so `_read` gets called over and over again automatically by the native Node Stream implementation. When `_read` gets called, it waits for the default number of bytes to get sent over from T2 before calling the callback with data.
```.js
USBConnection.prototype._read = function(n) {
  var self = this;

  if (self.closed) {
    return self.push(null);
  }

  self.epIn.transfer(4096, function(err, data) {
    if (err) {
      self.emit('error', err);
    } else {
      self.push(data);
    }
  });
};
```
Once the data is pushed, `_read` is called again and it starts waiting for more inbound USB data. That's why `usb` throws the error that a request is pending when the we try to close the connection.

We need a way to only have `_read` called when data is actually being waited on, or a way to cancel the request (is that possible @kevinmehall?)

**Comment by johnnyman727** [comment]
Perhaps I should be using the [USB Streams](https://github.com/kevinmehall/node-usb#startstreamntransfers3-transfersizemaxpacketsize) interface.

**Comment by johnnyman727** [comment]
Okay, I believe this was fixed. @kevinmehall can you check out https://github.com/tessel/t2-cli/commit/87b04cac813bf5c5795998b3c1b75cf3c676bf87 and make sure I didn't do anything stupid with USB?

File: test/unit/provision.js [hunk 2]
```diff
@@ -82,12 +114,11 @@ exports['controller.provisionTessel'] = {
     this.provisionTessel().catch(function(error) {
       test.equal(error, 'Keys already exist. Refusing to overwrite them.');
       test.done();
-    });
+    }.bind(this));
```
Comment (rwaldron) [inline test/unit/provision.js]: ?

**Comment by rwaldron** [comment]
:clap: 

```sh
rwaldron at nova in ~
$ rm -r .tessel/
rwaldron at nova in ~
$ t2 provision
INFO Connecting to Tessel...
INFO Connected to arnold over USB
INFO Creating public and private keys for Tessel authentication...
INFO SSH Keys written.
INFO Authenticating Tessel with public key...
INFO Tessel authenticated with public key.
rwaldron at nova in ~
$ t2 provision
WARN Keys already exist. Refusing to overwrite them.
rwaldron at nova in ~
$ t2 provision -f
INFO Connecting to Tessel...
INFO Connected to arnold over USB
INFO Creating public and private keys for Tessel authentication...
INFO SSH Keys written.
INFO Authenticating Tessel with public key...
INFO Tessel authenticated with public key.
```

**Pull request merged**